app/models/proxy.rb

```diff
@@ -529,49 +516,105 @@ def configuration_is_object
+      parsed_config = read_and_parse_policies_config(policies_config)
+
+      if parsed_config.detect { |c| c['name'] == DEFAULT_POLICY['name'] }
+        parsed_config
+      else
+        parsed_config.push(DEFAULT_POLICY)
+      end
+
+      @policies_config = parsed_config.map { |attrs| PolicyConfig.new(attrs) }
     end
 
     def self.name
       'PoliciesConfig'
     end
 
+    def to_json(*args)
+      map(&:to_h).to_json(*args)
+    end
+
+    def find_by(name:, version:)
+      find { |config| config.name == name && config.version == version }
+    end
+    alias find_policy_config_by find_by
+
+    def chain
+      select(&:enabled).map(&:to_h).map(&:stringify_keys).map { |p| p.slice('name', 'version', 'configuration') }
+    end
+    alias policy_chain chain
+
+    def ==(other)
+      self.class == other.class &&
+        policies_config == other.policies_config
+    end
+    alias eql? ==
+
     private
 
+    def read_and_parse_policies_config(attr_policies_config)
+      read_and_parse_policies_config!(attr_policies_config)
+    rescue JSON::ParserError
+      # TODO: is this a proper invalid value handling? In original
+      logger.error "Invalid attribute policies_config: #{attr_policies_config}"
+      []
+    end
+
+    def read_and_parse_policies_config!(attr_policies_config)
+      if attr_policies_config.blank?
+        []
+      elsif attr_policies_config.is_a? Array
+        attr_policies_config
+      else
+        Array(JSON.parse(attr_policies_config))
+      end
+    end
+
     def policies_configs_are_correct
-      policies_config.each do |policy_config|
+      each do |policy_config|
         # TODO: 5: errors.merge!(policy_config.errors)
         policy_config.errors.each { |attribute, message| errors.add(attribute, message) } unless policy_config.valid?
       end
+      unless detect { |c| c.name == DEFAULT_POLICY['name'] }
+        errors.add(:policies_config, "must include default policy")
+      end
     end
   end
 
-  def read_and_parse_policies_config
-    read_and_parse_policies_config!
-  rescue JSON::ParserError
-    []
-  end
+  class PoliciesConfigType < ActiveRecord::Type::Text
+    def cast(value)
+      if value.is_a?(PoliciesConfig)
+        value
+      else
+        PoliciesConfig.new(value)
+      end
+    end
 
-  def read_and_parse_policies_config!
-    attr_policies_config = read_attribute(:policies_config)
-    attr_policies_config.blank? ? [] : Array(JSON.parse(attr_policies_config))
+    def serialize(value)
+      value.to_json
+    end
+
+    def changed_in_place?(raw_old_value, new_value)
+      new_value != cast(raw_old_value)
+    end
   end
 
   def policies_config_structure
-    parsed_config = read_and_parse_policies_config!
-    policies_object = PoliciesConfig.new(parsed_config)
-    return if policies_object.valid?
-    policies_object.errors.each do |attribute, message|
+    return if policies_config.valid?
+    policies_config.errors.each do |attribute, message|
```

Comment:
At least my linter complaints about no adding blank lines after guards 😄 
```suggestion

    policies_config.errors.each do |attribute, message|
```